x-pack/plugins/infra/public/containers/metrics/with_metrics_time.tsx

```diff
@@ -4,89 +4,152 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import React from 'react';
-import { connect } from 'react-redux';
-import { createSelector } from 'reselect';
-
-import { metricTimeActions, metricTimeSelectors, State } from '../../store';
-import { asChildFunctionRenderer } from '../../utils/typed_react';
-import { bindPlainActionCreators } from '../../utils/typed_redux';
+import createContainer from 'constate-latest';
+import moment from 'moment';
+import React, { useContext, useEffect, useState } from 'react';
+import { InfraTimerangeInput } from '../../graphql/types';
+import { useInterval } from '../../hooks/use_interval';
 import { replaceStateKeyInQueryString, UrlStateContainer } from '../../utils/url_state';
 
-export const withMetricsTime = connect(
-  (state: State) => ({
-    currentTimeRange: metricTimeSelectors.selectRangeTime(state),
-    isAutoReloading: metricTimeSelectors.selectIsAutoReloading(state),
-    urlState: selectTimeUrlState(state),
-  }),
-  bindPlainActionCreators({
-    setRangeTime: metricTimeActions.setRangeTime,
-    startMetricsAutoReload: metricTimeActions.startMetricsAutoReload,
-    stopMetricsAutoReload: metricTimeActions.stopMetricsAutoReload,
-  })
-);
+interface MetricsTimeState {
+  timeRange: InfraTimerangeInput;
+  setTimeRange: (timeRange: InfraTimerangeInput) => void;
+  refreshInterval: number;
+  setRefreshInterval: (refreshInterval: number) => void;
+  isAutoReloading: boolean;
+  setAutoReload: (isAutoReloading: boolean) => void;
+}
+
+export const useMetricsTime = () => {
+  const [isAutoReloading, setAutoReload] = useState(false);
+  const [refreshInterval, setRefreshInterval] = useState(5000);
+  const [timeRange, setTimeRange] = useState({
+    from: moment()
+      .subtract(1, 'hour')
+      .valueOf(),
+    to: moment().valueOf(),
+    interval: '>=1m',
+  });
+
+  useInterval(
+    () => {
+      const range = timeRange.to - timeRange.from;
+      setTimeRange({
+        from: moment()
+          .subtract(range, 'ms')
+          .valueOf(),
+        to: moment().valueOf(),
+        interval: '>=1m',
+      });
+    },
+    isAutoReloading ? refreshInterval : null
+  );
+
+  useEffect(
+    () => {
+      if (isAutoReloading) {
+        const range = timeRange.to - timeRange.from;
+        setTimeRange({
+          from: moment()
+            .subtract(range, 'ms')
+            .valueOf(),
+          to: moment().valueOf(),
+          interval: '>=1m',
+        });
```

Comment:
How about we factor this block out as a `const setTimeRangeToNow = useCallback(..., [timerange.from, timerange.to])` and use it from both the interval and the effect?